drivers/can/can_stm32h7.c

```diff
@@ -0,0 +1,333 @@
+/*
+ * Copyright (c) 2022 Blue Clover
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <drivers/can.h>
+#include <drivers/clock_control/stm32_clock_control.h>
+#include <drivers/clock_control.h>
+#include <drivers/pinctrl.h>
+#include <kernel.h>
+#include <soc.h>
+#include <stm32_ll_rcc.h>
+#include "can_mcan.h"
+
+#include <logging/log.h>
+LOG_MODULE_DECLARE(can_driver, CONFIG_CAN_LOG_LEVEL);
+
+#define DT_DRV_COMPAT st_stm32h7_fdcan
+
+struct can_stm32h7_config {
+	struct can_mcan_msg_sram *msg_sram;
+	void (*config_irq)(void);
+	struct can_mcan_config mcan_cfg;
+	const struct pinctrl_dev_config *pcfg;
+	struct stm32_pclken pclken;
+};
+
+struct can_stm32h7_data {
+	struct can_mcan_data mcan_data;
+};
+
+int can_stm32h7_get_core_clock(const struct device *dev, uint32_t *rate)
+{
+	ARG_UNUSED(dev);
+	int rate_tmp;
+
+	rate_tmp = LL_RCC_GetFDCANClockFreq(LL_RCC_FDCAN_CLKSOURCE);
+
+	if (rate_tmp == LL_RCC_PERIPH_FREQUENCY_NO) {
+		LOG_ERR("Can't read core clock");
+		return -EIO;
+	}
+
+	*rate = rate_tmp;
+
+	LOG_DBG("rate=%d", *rate);
+
+	return 0;
+}
+
+int can_stm32h7_get_max_filters(const struct device *dev, enum can_ide id_type)
+{
+	if (id_type == CAN_STANDARD_IDENTIFIER) {
+		return NUM_STD_FILTER_DATA;
+	} else {
+		return NUM_EXT_FILTER_DATA;
+	}
+}
+
+int can_stm32h7_clock_enable(const struct device *dev)
+{
+	const struct can_stm32h7_config *cfg = dev->config;
+	const struct device *clk = DEVICE_DT_GET(STM32_CLOCK_CONTROL_NODE);
+
+	LL_RCC_SetFDCANClockSource(LL_RCC_FDCAN_CLKSOURCE_PLL1Q);
+
+	if (clock_control_on(clk,
+		(clock_control_subsys_t *) &cfg->pclken) != 0) {
+		LOG_ERR("failure enabling clock");
+		return -EIO;
+	}
+
+	if (!LL_RCC_PLL1Q_IsEnabled()) {
+		LOG_ERR("PLL1Q clock must be enabled!");
+		return -EIO;
+	}
+
+	return 0;
+}
+
+void can_stm32h7_set_state_change_cb(const struct device *dev,
+				     can_state_change_callback_t cb,
+				     void *user_data)
+{
+	struct can_stm32h7_data *data = dev->data;
+
+	data->mcan_data.state_change_cb = cb;
+	data->mcan_data.state_change_cb_data = user_data;
+}
+
+static int can_stm32h7_init(const struct device *dev)
+{
+	const struct can_stm32h7_config *cfg = dev->config;
+	const struct can_mcan_config *mcan_cfg = &cfg->mcan_cfg;
+	struct can_stm32h7_data *data = dev->data;
+	struct can_mcan_data *mcan_data = &data->mcan_data;
+	struct can_mcan_msg_sram *msg_ram = cfg->msg_sram;
+	int ret;
+
+	/* Configure dt provided device signals when available */
+	ret = pinctrl_apply_state(cfg->pcfg, PINCTRL_STATE_DEFAULT);
+	if (ret < 0) {
+		LOG_ERR("CAN pinctrl setup failed (%d)", ret);
+		return ret;
+	}
+
+	ret = can_stm32h7_clock_enable(dev);
+	if (ret) {
+		return ret;
+	}
+
+	ret = can_mcan_init(dev, mcan_cfg, msg_ram, mcan_data);
+	if (ret) {
```

Comment:
please be consistent: better use essentially boolean expressions (`ret < 0`)

Reply:
Made consistent.